File: haskell/17387_intersected_line_segments_2.hs
import Control.Monad (replicateM)

data Point = Point Int Int

instance Eq Point where
  (Point x1 y1) == (Point x2 y2) = x1 == x2 && y1 == y2

instance Ord Point where
  (Point x1 y1) <= (Point x2 y2) = if x1 == x2 then y1 <= y2 else x1 <= x2

data Line = Line Point Point

xOf :: Point -> Int
xOf (Point x _) = x

yOf :: Point -> Int
yOf (Point _ y) = y

fstPoint :: Line -> Point
fstPoint (Line p _) = p

sndPoint :: Line -> Point
sndPoint (Line _ p) = p

ccw :: Point -> Point -> Point -> Int
ccw p1 p2 p3
  | cp == 0 = 0
  | cp > 0 = 1
  | otherwise = -1
  where
    a = (xOf p2 - xOf p1, yOf p2 - yOf p1)
    b = (xOf p3 - xOf p1, yOf p3 - yOf p1)
    cp = fst a * snd b - snd a * fst b

main :: IO ()
main = do
  list@(l1 : l2 : _) <- replicateM 2 $ do
    input <- getLine
    let (p : q : r : s : _) = read <$> words input :: [Int]
        p1 = Point p q
        p2 = Point r s
    return $ Line (min p1 p2) (max p1 p2)

  let onL1 = ccw (fstPoint l1) (sndPoint l1) (fstPoint l2) * ccw (fstPoint l1) (sndPoint l1) (sndPoint l2)
      onL2 = ccw (fstPoint l2) (sndPoint l2) (fstPoint l1) * ccw (fstPoint l2) (sndPoint l2) (sndPoint l1)

  if onL1 <= 0 && onL2 <= 0
    then 
      if onL1 == 0 && onL2 == 0 
        then 
          if fstPoint l1 <= sndPoint l2 && fstPoint l2 <= sndPoint l1 
            then print 1 
            else print 0
        else print 1
    else print 0

File: rust/src/bin/17387_intersected_line_segments_2.rs
use std::{io, cmp::{Ordering, self}};

use scanner::UnsafeScanner;

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
struct Point { x: i64, y: i64 }

impl Point {
    fn new(x: i64, y: i64) -> Self {
        Self { x, y }
    }
}

struct Line { p1: Point, p2: Point }

impl Line {
    fn new(p1: Point, p2: Point) -> Self {
        Self { p1, p2 }
    }
}

fn ccw(p1: &Point, p2: &Point, p3: &Point) -> i32 {
    // Defines a to be a vector from p1 to p2.
    // Similar for b, but from p1 to p3.
    let a = (p2.x - p1.x, p2.y - p1.y);
    let b = (p3.x - p1.x, p3.y - p1.y);

    let cp = a.0 * b.1 - a.1 * b.0;
    match cp.cmp(&0) {
        Ordering::Less  => -1,
        Ordering::Equal => 0,
        Ordering::Greater => 1,
    }
}

fn main() {
    let stdin = io::stdin();
    let mut scan = UnsafeScanner::new(stdin.lock());

    let mut ls = Vec::with_capacity(2);
    (0..2).for_each(|_| {
        let p1 = Point::new(scan.token::<i64>(), scan.token::<i64>());
        let p2 = Point::new(scan.token::<i64>(), scan.token::<i64>());
        ls.push(Line::new(cmp::min(p1, p2), cmp::max(p1, p2)));
    });

    let on_l1 = ccw(&ls[0].p1, &ls[0].p2, &ls[1].p1) * ccw(&ls[0].p1, &ls[0].p2, &ls[1].p2);
    let on_l2 = ccw(&ls[1].p1, &ls[1].p2, &ls[0].p1) * ccw(&ls[1].p1, &ls[1].p2, &ls[0].p2);

    if on_l1 <= 0 && on_l2 <= 0 {
        if on_l1 == 0 && on_l2 == 0 {
            if ls[0].p1 <= ls[1].p2 && ls[1].p1 <= ls[0].p2 { println!("1"); }
            else { println!("0"); }
            return;
        }
        println!("1");
    }
    else { println!("0"); }
}

mod scanner {
    use std::{io, str};

    pub(crate) struct UnsafeScanner<R> {
        reader: R,
        buf_str: Vec<u8>,
        buf_iter: str::SplitAsciiWhitespace<'static>,
    }

    impl <R: io::BufRead> UnsafeScanner<R> {
        pub(crate) fn new(reader: R) -> Self {
            Self {
                reader,
                buf_str: vec![],
                buf_iter: "".split_ascii_whitespace(),
            }
        }

        pub(crate) fn token<T: str::FromStr>(&mut self) -> T {
            loop {
                if let Some(token) = self.buf_iter.next() {
                    return token.parse().ok().expect("Failed parse");
                }
                self.buf_str.clear();
                self.reader
                    .read_until(b'\n', &mut self.buf_str)
                    .expect("Failed read");
                self.buf_iter = unsafe {
                    let slice = str::from_utf8_unchecked(&self.buf_str);
                    std::mem::transmute(slice.split_ascii_whitespace())
                }
            }
        }
    }
}
